DNX init.ps1

deepakaravindr opened the pull request:
1. Implemented the IVsScriptExecutor interface
2. ScriptExector does the actual execution of init.ps1 scripts. It tries to
   locate the package on packages folder. If not found, it tries to locate the
   package in global packages folder
3. ScriptExecutor has been refactored a little bit to only execute init.ps1
   that has not been executed already. It also caches solution directory.
4. On solution open though, PowershellHost calls into IScriptExecutor to
   reset the cached solution directory and the cache of successful executions
   of init.ps1. Meaning, init.ps1 will get executed on every solution open.
5. Added a couple of functional tests to test the new service IVsScriptExecutor
   both on a regular class library and a DNX project
6. Ran all the functional tests both existing and new and ensures that init.ps1 scenarios are covered.

@yishaigalatzer @emgarten

deepakaravindr committed:
Renamed VSScriptExecutor to be ScriptExecutor which is under the
NuGetConsole namespace, since, the VS prefix should only be used by the
implementations of IVS interfaces in NuGet VS Extensibility.

deepakaravindr committed:
Interface definition of IVsScriptExecutor and stub implementation
of VsScriptExecutor

deepakaravindr committed:
1. Implemented the IVsScriptExecutor interface
2. ScriptExector does the actual execution of init.ps1 scripts. It tries to
locate the package on packages folder. If not found, it tries to locate the
package in global packages folder
3. ScriptExecutor has been refactored a little bit to only execute init.ps1
that has not been executed already. It also caches solution directory.
4. On solution open though, PowershellHost calls into IScriptExecutor to
reset the cached solution directory and the cache of successful executions
of init.ps1. Meaning, init.ps1 will get executed on every solution open.
5. Added a couple of functional tests to test the new service IVsScriptExecutor
both on a regular class library and a DNX project

emgarten commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsScriptExecutor.cs:
```diff
@@ -0,0 +1,46 @@
+                throw new ArgumentNullException(CommonResources.Argument_Cannot_Be_Null_Or_Empty, nameof(packageId));
+            }
+
+            if (string.IsNullOrEmpty(packageId))
```
packageVersion

deepakaravindr replied:
Done

emgarten commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsScriptExecutor.cs:
```diff
@@ -0,0 +1,46 @@
+                    nameof(packageVersion));
+            }
+
+            var version = new NuGetVersion(packageVersion);
```
Does this need special error handling? The string may not be a version.

deepakaravindr replied:
I believe NuGetVersion throws the right messages. I can probably add some functional tests for these cases.

yishaigalatzer replied:
please do

yishaigalatzer replied:
Actually you want to throw an argument exception here if the version is bad

emgarten commented on src/NuGet.Clients/VisualStudio/Extensibility/IVsScriptExecutor.cs:
```diff
@@ -0,0 +1,28 @@
+namespace NuGet.VisualStudio
+{
+    /// <summary>
+    /// Contains methods to execute powershell scripts from package(s) in a solution
```
'Contains methods' isn't needed here

deepakaravindr replied:
Done

emgarten commented on src/NuGet.Clients/VisualStudio/Extensibility/IVsScriptExecutor.cs:
```diff
@@ -0,0 +1,28 @@
+        /// Executes the init script of the given package if available.
+        /// 1) If the init.ps1 script has already been executed by the powershell host, it will not be executed again.
+        /// True is returned.
+        /// 2) If the package is found on the packages folder, it will be used. Otherwise, the one on global packages
```
_found in_ instead of _found on_. 
_in global_ instead of _on global_

deepakaravindr replied:
Done.

emgarten commented on src/NuGet.Clients/VisualStudio/Extensibility/IVsScriptExecutor.cs:
```diff
@@ -0,0 +1,28 @@
+        /// <param name="packageId">Id of the package whose init.ps1 will be executed.</param>
+        /// <param name="packageVersion">Version of the package whose init.ps1 will be executed.</param>
+        /// <returns>Returns true if the script was executed or has been executed already.</returns>
+        /// <remarks>This method throws if the init.ps1 being executed throws </remarks>
```
End with .

deepakaravindr replied:
Done

emgarten commented on src/NuGet.Clients/VsConsole/PowerShellHost/PowerShellHost.cs:
```diff
@@ -396,7 +400,22 @@ private void UpdateWorkingDirectory()
-                                    Runspace.ExecuteScript(pathToPackage, scriptPath, package);
+                                    if (File.Exists(scriptPath))
+                                    {
+                                        if (_scriptExecutor.TryMarkVisited(package, true))
```
name parameter on the bool here and down below

deepakaravindr replied:
Done

emgarten commented on test/TestExtensions/API.Test/InternalAPITestHook.cs:
```diff
@@ -57,5 +58,21 @@ public static void RestorePackageApi()
+            // Calling thread is powershell execution thread and ExecuteInitScriptAsync needs to switch to
+            // Powershell execution thread to execute the scripts
+            var task = Task.Run(async () => await scriptExecutor.ExecuteInitScriptAsync(id, version));
+            Task.WaitAny(task, Task.Delay(1000));
```
1s seems low for a test

deepakaravindr replied:
I have increased it to 3 secs

emgarten commented on test/EndToEnd/tests/ServicesTest.ps1:
```diff
@@ -485,5 +485,46 @@ function Test-InstallPackageAPIBindingRedirect
+    Assert-True ($global:PackageInitPS1Var -eq 1)
+}
 
+function Test-ExecuteInitPS1OnProjectK
```
Test-ExecuteInitPS1OnAspNetCore :smile:

deepakaravindr replied:
Done

emgarten commented on src/NuGet.Clients/VsConsole/Console/ScriptExecutor.cs:
```diff
@@ -0,0 +1,328 @@
+                    nuGetProject.TryGetMetadata(NuGetProjectMetadataKeys.TargetFramework, out targetFramework);
+
+                    // targetFramework can be null for unknown project types
+                    string shortFramework = targetFramework == null ? string.Empty : targetFramework.GetShortFolderName();
```
nit: slightly shorter: targetFramework?.GetShortFolderName() ?? string.Empty

deepakaravindr replied:
I have fixed it. But, this is not a new change

emgarten commented:
:shipit:

yishaigalatzer commented on src/NuGet.Clients/VisualStudio/Extensibility/IVsScriptExecutor.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace NuGet.VisualStudio
+{
+    /// <summary>
```
Add - For internal use only

deepakaravindr replied:
Added 'Intended for internal use only.' on the remarks section on the interface

yishaigalatzer commented on src/NuGet.Clients/VsConsole/Console/ScriptExecutor.cs:
```diff
@@ -0,0 +1,328 @@
+
+        public void Reset()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
```
If you are requiring this on the constructor (which seems like you do), Add a call on it, so a user gets a good callstack for the exception if this was imported from a background thread.

deepakaravindr replied:
Fixed. As discussed in the next comment, I have removed the check altogether. So, not needed from the constructor either.

yishaigalatzer commented on src/NuGet.Clients/VsConsole/Console/ScriptExecutor.cs:
```diff
@@ -0,0 +1,328 @@
+
+            InitScriptExecutions.Clear();
+
+            foreach (var item in packagesWithoutInitPS1)
```
I'm not clear why all of this is running on the main thread.

deepakaravindr replied:
I have changed it. Properties on SolutionManager such as IsSolutionAvailable and SolutionDirectory already switch to the main thread if not already on main thread. So, removed the ThrowIfNotOnUIThread.

yishaigalatzer commented on src/NuGet.Clients/VsConsole/Console/ScriptExecutor.cs:
```diff
@@ -0,0 +1,328 @@
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (SolutionManager.IsSolutionAvailable)
```
This is really the only thing that happens on the constructor, so please decouple from reset (perhaps reset calls this)

deepakaravindr replied:
As mentioned in a subsequent comment, PowershellHost uses Reset and not just the constructor

yishaigalatzer commented on src/NuGet.Clients/VsConsole/Console/ScriptExecutor.cs:
```diff
@@ -0,0 +1,328 @@
+        [Import]
+        public IOutputConsoleProvider OutputConsoleProvider { get; set; }
+
+        public void Reset()
```
reading this again, I don't follow what is the intent of this method.

deepakaravindr replied:
This method is used by powershellhost to reset the script executor during solution open before running any init.ps1 scripts

yishaigalatzer replied:
Lets talk about this, reset does something very specific, it looks like it keeps a list of only init.ps1 that didn't run. It seems odd, and I don't quite understand what it is for. So I'd like to understand the algorithm, the purpose etc. in comments or see if it needs to stay at all in its current form.

yishaigalatzer commented on src/NuGet.Clients/VsConsole/Console/ScriptExecutor.cs:
```diff
@@ -0,0 +1,328 @@
+            string effectiveGlobalPackagesFolder = null;
+            if (!string.IsNullOrEmpty(SolutionDirectory))
+            {
+                var packagesFolder = PackagesFolderPathUtility.GetPackagesFolderPath(SolutionDirectory, Settings);
```
resolving packages folder doesn't seem to belong here. There should only be one packages folder for this scenario, which is the global packages folder. This interface should be limited to that scenario and clearly document it

deepakaravindr replied:
Fixed. I have refactored and removed unnecessary code

yishaigalatzer commented on src/NuGet.Clients/VisualStudio/Extensibility/IVsScriptExecutor.cs:
```diff
@@ -0,0 +1,28 @@
+    /// </summary>
+    [ComImport]
+    [Guid("8c6750b0-214f-4052-93b2-284f12ebb896")]
+    public interface IVsScriptExecutor
```
IVsGlobalPackagesInitScriptExecutor

deepakaravindr replied:
Should I call it IVsGlobalPackagesScriptExecutor? Because, method is called ExecuteInitScriptAsync. (For now, I have gone with your suggestion)

deepakaravindr committed:
1. Interface has been renamed to VsGlobalPackagesInitScriptExecutor and is
only applicable for global packages
2. Incorporated other comments

deepakaravindr commented:
I have incorporated most comments. Need to fix a couple of more comments.

yishaigalatzer commented:
Thanks, I'll wait for your final iteration

deepakaravindr committed:
Added a unit test library for VSGlobalPackagesInitScriptExecutor to test
the negative inputs

deepakaravindr commented:
I needed to add a few unit tests for VsGlobalPackagesInitScriptExecutor. Added them. It is ready

yishaigalatzer commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsGlobalPackagesInitScriptExecutor.cs:
```diff
@@ -0,0 +1,46 @@
+        {
+            if (string.IsNullOrEmpty(packageId))
+            {
+                throw new ArgumentNullException(CommonResources.Argument_Cannot_Be_Null_Or_Empty, nameof(packageId));
```
use ArgumentException

yishaigalatzer commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsGlobalPackagesInitScriptExecutor.cs:
```diff
@@ -0,0 +1,46 @@
+
+            if (string.IsNullOrEmpty(packageVersion))
+            {
+                throw new ArgumentNullException(
```
same as above

yishaigalatzer commented on test/TestExtensions/API.Test/InternalAPITestHook.cs:
```diff
@@ -57,5 +58,21 @@ public static void RestorePackageApi()
+            // Calling thread is powershell execution thread and ExecuteInitScriptAsync needs to switch to
+            // Powershell execution thread to execute the scripts
+            var task = Task.Run(async () => await scriptExecutor.ExecuteInitScriptAsync(id, version));
+            Task.WaitAny(task, Task.Delay(3000));
```
What is this delay for?

yishaigalatzer commented:
:shipit: when feedback is addressed, mostly I'm looking for making reset just a call to clear

deepakaravindr committed:
1. The cache tracking init ps1 executions uses an enum to track the state
instead of using true or false. Made changes corresponding to that.
2. Also, on ScriptExecutor reset, the entire cache is cleared.
3. Changed ArgNull to ArgException on string null or empty. Fixed and added
tests

deepakaravindr commented:
Merged with https://github.com/NuGet/NuGet.Client/commit/6c17c6ed900befeee2956c5d6a7661b5824c466e
